File: 2019/Elixir/3/main.exs
require File
require IO

defmodule Main do
    # Part 1
    def main(filePath \\ "input.txt") do
        case File.read filePath do
            {:ok, contents} ->
                    [firstWirePath, secondWirePath] = contents 
                        |> String.split("\n", trim: true)
                        |> fn l -> Enum.map(l, fn i -> String.split(i, ",", trim: true) end) end.()
                        |> fn l -> Enum.map(l, fn i -> Enum.map(i, &WirePath.fromString/1) end) end.()
                    
                    [firstWireSegments, secondWireSegments] = 
                        [firstWirePath, secondWirePath] |> Enum.map(&wirePathsToSegments/1)

                    intersections = findSegmentsIntersections(firstWireSegments, secondWireSegments) 
                        |> Enum.drop(1) # wires always Intersec at the origin, do drop it
                    
                    intersections
                        |> Enum.map(fn t -> {t , taxyCabDistanceFromOrigin(t)} end)
                        |> Enum.min_by(fn {_, distance} -> distance end)

                    lengthsToIntersections = intersections |> Enum.map(fn i -> 
                        # If a wire visits a position on the grid multiple times, use the steps value from the first time it visits that position
                        w1PathWithOutEndSegment = Enum.take_while(firstWireSegments, fn s -> not crossesPoint(i, s) end) 
                        w1PathLastSegment = Enum.drop(firstWireSegments, Kernel.length(w1PathWithOutEndSegment)) |> Kernel.hd
                        w1LastSteps = stepsForSegmentToPoint(w1PathLastSegment, i)
                        w2PathWithOutEndSegment = Enum.take_while(secondWireSegments, fn s -> not crossesPoint(i, s) end)
                        w2PathLastSegment = Enum.drop(secondWireSegments, Kernel.length(w2PathWithOutEndSegment)) |> Kernel.hd
                        w2LastSteps = stepsForSegmentToPoint(w2PathLastSegment, i)
                        {i, { (w1PathWithOutEndSegment |> (Enum.map(&stepsForSegment/1) |> Enum.sum)) + w1LastSteps, w1PathWithOutEndSegment }, { (w2PathWithOutEndSegment |> (Enum.map(&stepsForSegment/1) |> Enum.sum)) + w2LastSteps, w2PathWithOutEndSegment }} end)

                    lengthsToIntersections
                        |> Enum.min_by(fn {_, {l1, _}, {l2, _}} -> l1 + l2 end) 
                        |> fn {_, {l1, _}, {l2, _}} -> l1 + l2 end.()
                
            {:error, reason} ->
                IO.puts("Cannot open input.txt: #{reason}")
        end
    end

    def taxyCabDistanceFromOrigin({x, y}) do
        Kernel.abs(x) + Kernel.abs(y)
    end

    def findSegmentsIntersections(firstWireSegments, secondWireSegments) do

        allIntersections = 
            (for segment1 <- firstWireSegments, segment2 <- secondWireSegments,  into: [], do: intersections(segment1, segment2)) 
                |> Enum.filter(fn x -> not (Enum.empty? x) end)
        Enum.concat(allIntersections)
    end

    def test() do
        IO.inspect(main("example.txt"))
        IO.inspect(main("test1.txt"))
        IO.inspect(main("test2.txt"))
    end


    def wirePathsToSegments(wirePaths) do    
        origin = {0, 0}
        [head | tail] = wirePaths
        accumutalor = [{ origin, applyWirePathDisplacement(origin, head) }]

        Enum.reduce(tail, accumutalor, fn element, acc -> 
            {_, endOfSegment} = List.last(acc)
            acc ++ [{endOfSegment, applyWirePathDisplacement(endOfSegment, element) }] end
        )
    end

    def applyWirePathDisplacement(point, wirePath) do
        {x, y} = point
        case wirePath.direction do
            :R -> { x + wirePath.amount, y }
            :L -> { x - wirePath.amount, y }
            :U -> { x, y + wirePath.amount }
            :D -> { x, y - wirePath.amount }
        end
    end

    def intersections(segment1, segment2) do
        { { s1StartX, s1StartY } , { s1EndX, s1EndY} } = segment1
        { { s2StartX, s2StartY } , { s2EndX, s2EndY} } = segment2

        s1XSet = Enum.into(s1StartX..s1EndX, MapSet.new)
        s2XSet = Enum.into(s2StartX..s2EndX, MapSet.new)

        s1YSet = Enum.into(s1StartY..s1EndY, MapSet.new) 
        s2YSet = Enum.into(s2StartY..s2EndY, MapSet.new) 

        xIntersections = MapSet.intersection(s1XSet, s2XSet) |> MapSet.to_list
        yIntersections = MapSet.intersection(s1YSet, s2YSet) |> MapSet.to_list

        cond do
            Enum.empty?(xIntersections) or Enum.empty?(yIntersections) -> []
            Enum.count(xIntersections) > 1 -> Enum.into(xIntersections, [], fn x -> {x, List.first(yIntersections) } end)
            Enum.count(yIntersections) > 1 -> Enum.into(yIntersections, [], fn y -> {List.first(xIntersections), y} end)
            Enum.count(xIntersections) == 1 and Enum.count(yIntersections) == 1 -> [{List.first(xIntersections), List.first(yIntersections)}]
        end
    end

    def crossesPoint({x, y}, { {p1x, p1y}, {p2x, p2y} }) do

        crossesX = Enum.member?(p1x..p2x, x) 
        crossesY = Enum.member?(p1y..p2y, y)

        crossesX and crossesY
    end

    def stepsForSegment({{p1x, p1y}, {p2x, p2y}}) do
        Kernel.abs(p1x-p2x) + Kernel.abs(p1y-p2y)
    end

    def stepsForSegmentToPoint({{p1x, p1y}, _}, {x, y}) do
        Kernel.abs(p1x-x) + Kernel.abs(p1y-y)
    end
end


defmodule WirePath do
    defstruct direction: :R, amount: 0 

    def fromString(string) do
        << d :: utf8, rest:: binary>> = string
        %WirePath { direction: List.to_string([d]) |> String.to_atom, amount: List.to_string([rest]) |> String.to_integer }
    end
end

File: 2019/Haskell/3/main.hs
#!/usr/bin/env stack
-- stack --system-ghc --resolver lts-14.16 script

{-# LANGUAGE OverloadedStrings #-}

import           Control.Applicative                      ( (<$>) )

import           Data.Char                                ( isDigit
                                                          , toUpper
                                                          )
import qualified Text.ParserCombinators.ReadP  as P
import qualified Text.Read                     as R
                                                          ( readP_to_Prec )
import           Data.List
import qualified Data.Text                     as Text
import qualified Data.Text.IO                  as Text
import qualified Data.Set                      as Set
import qualified Text.Read                     as R
import qualified Text.Read.Lex                 as R

import           Debug.Trace

data Direction = R | L | U | D deriving (Show, Eq, Read)
data WirePath = WirePath { direction :: Direction,  amount :: Integer} deriving (Show, Eq)

-- All credit for this instace to @jaspervdj https://gist.github.com/jaspervdj/86a0179f44e083234d4c2377103a357a
instance R.Read WirePath where
  readPrec = R.readP_to_Prec $ const $ do
    c <- P.get
    s <- P.munch1 isDigit
    return $ WirePath { direction = read [toUpper c], amount = read s }

data Point = Point { x :: Integer, y :: Integer } deriving (Show, Eq)
data Segment = Segment { p1 :: Point, p2 :: Point } deriving (Show, Eq)

main :: IO ()
main = test --solvePathsFromFile "example.txt"


solvePathsFromFile :: String -> IO ()
solvePathsFromFile file = do
  withPaths@[wirePath1, wirePath2] <-
    (map (map (read . Text.unpack)) :: [[Text.Text]] -> [[WirePath]])
    .   map (Text.splitOn ",")
    .   Text.lines
    <$> (Text.readFile file)
  let [segs1, segs2] = wirePathsToSegments <$> withPaths
  let crossings = map (\i -> (i, taxiCabDistanceFromOrigin i))
                      (tail . concat $ intersections <$> segs1 <*> segs2)

  let
    lengthsToIntersections = map
      (\intersection ->
        let
          w1PathWithOutEndSegment =
            segs1 `pathWithoutEndSegmentBefore` fst intersection
          w1PathLastSegment =
            head $ drop (length w1PathWithOutEndSegment) segs1
          w1LastSteps = w1PathLastSegment `stepsToPoint` fst intersection
          w2PathWithOutEndSegment =
            segs2 `pathWithoutEndSegmentBefore` fst intersection
          w2PathLastSegment =
            head $ drop (length w2PathWithOutEndSegment) segs2
          w2LastSteps = w2PathLastSegment `stepsToPoint` fst intersection
        in
          ( intersection
          , sum (stepsForSegment <$> w1PathWithOutEndSegment) + w1LastSteps
          , sum (stepsForSegment <$> w2PathWithOutEndSegment) + w2LastSteps
          )
      )
      crossings

  print
    $ ((\(_, dx, dy) -> dx + dy) :: ((Point, Integer), Integer, Integer)
        -> Integer
      )
        (minInList lengthsToIntersections)

 where
  pathWithoutEndSegmentBefore segments point =
    takeWhile (\s -> not $ s `crossesPoint` point) segments

applyWirePathDisplacementToPoint :: Point -> WirePath -> Point
applyWirePathDisplacementToPoint (Point x y) (WirePath direction amount) =
  case direction of
    R -> Point (x + amount) y
    L -> Point (x - amount) y
    U -> Point x (y + amount)
    D -> Point x (y - amount)

wirePathsToSegments :: [WirePath] -> [Segment]
wirePathsToSegments (p : ps) = reverse $ foldl
  (\acc element ->
    let (Segment _ endPoint) = head acc
    in  Segment endPoint (applyWirePathDisplacementToPoint endPoint element)
          : acc
  )
  [Segment (Point 0 0) (applyWirePathDisplacementToPoint (Point 0 0) p)]
  ps

intersections :: Segment -> Segment -> [Point]
intersections s1@(Segment (Point s1StartX s1StartY) (Point s1EndX s1EndY)) s2@(Segment (Point s2StartX s2StartY) (Point s2EndX s2EndY))
  = case (xIntersections, yIntersections) of
    (_  , [] ) -> []
    ([] , _  ) -> []
    (xs , [y]) -> foldl (\acc x -> Point x y : acc) [] xs
    ([x], ys ) -> foldl (\acc y -> Point x y : acc) [] ys
    a          -> error $ "Inputs don't follow challenge spec: " ++ show a
 where
  s1XSet         = Set.fromList [min s1StartX s1EndX .. max s1StartX s1EndX]
  s1YSet         = Set.fromList [min s1StartY s1EndY .. max s1StartY s1EndY]
  s2XSet         = Set.fromList [min s2StartX s2EndX .. max s2StartX s2EndX]
  s2YSet         = Set.fromList [min s2StartY s2EndY .. max s2StartY s2EndY]
  xIntersections = Set.toList $ s1XSet `Set.intersection` s2XSet
  yIntersections = Set.toList $ s1YSet `Set.intersection` s2YSet

taxiCabDistanceFromOrigin :: Point -> Integer
taxiCabDistanceFromOrigin (Point x y) = abs x + abs y

minInList
  :: [((Point, Integer), Integer, Integer)]
  -> ((Point, Integer), Integer, Integer)
minInList (triplet@(_, distanceX, distanceY) : rest) = findMin triplet rest
 where
  findMin currentMin@(_, minDistanceX, minDistanceY) (triplet@(_, distanceX, distanceY) : rest)
    | distanceX + distanceY >= minDistanceX + minDistanceY
    = findMin currentMin rest
    | otherwise
    = findMin triplet rest
  findMin currentMin [] = currentMin
minInList [] = error "Inputs don't follow challenge spec."

crossesPoint :: Segment -> Point -> Bool
crossesPoint s@(Segment (Point startX startY) (Point endX endY)) (Point x y) =
  let xs = Set.fromList [min startX endX .. max startX endX]
      ys = Set.fromList [min startY endY .. max startY endY]
  in  x `Set.member` xs && y `Set.member` ys

stepsToPoint :: Segment -> Point -> Integer
stepsToPoint s@(Segment (Point startX startY) _) (Point x y) =
  abs (startX - x) + abs (startY - y)

stepsForSegment :: Segment -> Integer
stepsForSegment s1@(Segment (Point startX startY) (Point endX endY)) =
  abs (startX - endX) + abs (startY - endY)

test :: IO ()
test =
  solvePathsFromFile "example.txt"
    <* solvePathsFromFile "test1.txt"
    <* solvePathsFromFile "test2.txt"
    <* solvePathsFromFile "input.txt"

